internal/sync/lifecycle.go

```diff
@@ -70,6 +76,37 @@ type LifecycleOnce struct {
 	state    atomic.Int32
 	startErr error
 	stopErr  error
+	started  chan struct{}
+}
+
+// WaitForStart blocks until the lifecycle has started, or the context expires.
+// Returns context.DeadlineExceeded if the context expires first.
+// Returns sync.AlreadyStopped if the lifecycle has already stopped.
+func (l *LifecycleOnce) WaitForStart(ctx context.Context) error {
+	state := LifecycleState(l.state.Load())
+	// Fast success in the common case
+	if state == Running {
+		return nil
+	}
+	// Fast fail if we ran already
+	if state > Running {
+		return ErrAlreadyStopped
+	}
+
+	// Create a started channel if none exists. Do so in a lock so we're not
+	// racing Start to set up for blocking.
+	l.lock.Lock()
+	if l.started == nil {
+		l.started = make(chan struct{}, 0)
+	}
+	l.lock.Unlock()
```

Comment:
I'm not sure this will quite work right. Imagine two goroutines A and B:

```
A: Start()
A: lock()
B: WaitForStart()
B: Waiting on lock
A: started == nil; move on
A: unlock()
B: lock()
B: started = make(chan)
B: unlock()
B: wait on started forever
```